src/lib/ledger_catchup/normal_catchup.ml

```diff
@@ -566,59 +649,72 @@ let run ~logger ~precomputed_values ~trust_system ~verifier ~network ~frontier
                         ~network ~frontier ~peers:random_peers ~target_hash
-                        ~job
+                        ~job ~blockchain_length_of_target_hash
                     with
                     | Ok (peer, hashes) ->
                         return (Ok (peer, hashes))
-                    | Error err ->
+                    | Error errors ->
                         [%log info]
                           "Could not download state hashes using random peers"
-                          ~metadata:[("error", Error_json.error_to_yojson err)] ;
-                        let transitions =
-                          List.concat_map subtrees ~f:Rose_tree.flatten
-                        in
-                        List.iter transitions ~f:(fun cached_transition ->
-                            let transition =
-                              Cached.peek cached_transition
-                              |> Envelope.Incoming.data
-                            in
-                            let transition_hash =
-                              External_transition.Initial_validated.state_hash
-                                transition
-                            in
-                            [%log error]
-                              ~metadata:
-                                [ ( "state_hash"
-                                  , State_hash.to_yojson transition_hash )
-                                ; ( "reason"
-                                  , `String
-                                      "no common ancestor with current root" )
-                                ; ( "protocol_state"
-                                  , External_transition.Initial_validated
-                                    .protocol_state transition
-                                    |> Mina_state.Protocol_state
-                                       .value_to_yojson ) ]
-                              "Validation error: external transition with \
-                               state hash $state_hash was rejected for reason \
-                               $reason" ) ;
-                        Mina_metrics.(
-                          Counter.inc Rejected_blocks.no_common_ancestor
-                            (Float.of_int @@ List.length transitions)) ;
-                        return (Error err) )
+                          ~metadata:
+                            [ ( "errors"
+                              , `List
+                                  (List.map errors ~f:(fun err ->
+                                       `String (display_error err) )) ) ] ;
+                        if contains_no_common_ancestor errors then (
+                          let transitions =
+                            List.concat_map subtrees ~f:Rose_tree.flatten
+                          in
+                          List.iter transitions ~f:(fun cached_transition ->
+                              let transition =
```

Comment:
Could you add make this one error message? Also, are the blocks received during catchup logged elsewhere? If so, just add the list of state hashes to the error message. Maybe we just need the best-tip we are trying to catchup to in the error message and the chain can be state-hashes